src/auth/discord.rs

```diff
@@ -1 +1,207 @@
+    let redirect_url = format!(
+        "{}://{}{}",
+        req.connection_info().scheme(),
+        req.connection_info().host(),
+        CALLBACK_URL
+    );
+    let discord_auth =
+        apis::discord_token_api(TOKEN_URL, redirect_url.as_str(), query.code.as_str()).await;
+    match discord_auth {
+        Ok(access_token) => {
+            let user_name = apis::user_details(&access_token).await?;
+            let user_detail_obj: UserDetail = UserDetail {
+                token: access_token.to_owned(),
+                user_name,
+            };
+            let user_detail_str = serde_json::to_string(&user_detail_obj)?;
+
+            return Ok(actix_web::HttpResponse::Found()
+                .cookie(
+                    actix_web::cookie::Cookie::build(
+                        fpm::auth::AuthProviders::Discord.as_str(),
+                        mc_obj
+                            .encrypt_to_base64(&user_detail_str)
+                            .as_str()
+                            .to_owned(),
+                    )
+                    .domain(fpm::auth::utils::domain(req.connection_info().host()))
+                    .path("/")
+                    .permanent()
+                    .finish(),
+                )
+                .append_header((actix_web::http::header::LOCATION, "/".to_string()))
+                .finish());
+        }
+        Err(err) => Ok(actix_web::HttpResponse::InternalServerError().body(err.to_string())),
+    }
+}
+// it returns identities which matches to given input
+pub async fn matched_identities(
+    _ud: UserDetail,
+    _identities: &[fpm::user_group::UserIdentity],
+) -> fpm::Result<Vec<fpm::user_group::UserIdentity>> {
+    /*let discord_identities = identities
+        .iter()
+        .filter(|identity| identity.key.starts_with("discord"))
+        .collect::<Vec<&fpm::user_group::UserIdentity>>();
+
+    if discord_identities.is_empty() {
+        return Ok(vec![]);
+    }*/
+
+    let matched_identities = vec![];
+
+    Ok(matched_identities)
+}
+
+pub mod apis {
+    #[derive(Debug, serde::Deserialize)]
+    pub struct DiscordAuthResp {
+        pub access_token: String,
+    }
+    // TODO: API to starred a repo on behalf of the user
+    // API Docs: https://discord.com/developers/docs/getting-started
+    //API EndPoints: https://github.com/GregTCLTK/Discord-Api-Endpoints/blob/master/Endpoints.md
+    // TODO: It can be stored in the request cookies
+    pub async fn user_details(token: &str) -> fpm::Result<String> {
+        // API Docs: https://discord.com/api/users/@me
+        // TODO: Handle paginated response
+        #[derive(Debug, serde::Deserialize)]
+        struct UserDetails {
+            username: String,
+        }
+        let user_obj: UserDetails = get_api("https://discord.com/api/users/@me", token).await?;
+
+        Ok(String::from(&user_obj.username))
+    }
+    pub async fn get_api<T: serde::de::DeserializeOwned>(url: &str, token: &str) -> fpm::Result<T> {
```

Comment:
This API is common for all the auth providers, so we can move it to the utility.